### DISCUSSION TITLE
is glsp possilble to generate specific code

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-glsp/glsp#1115] [Q&A]
there is a requirement when user performs a change in the diagram(my requirement is industrial Programming Language LD(Ladder Diagram)) like drag a position of Gnode(or add a Gnode), the client sends a notification to the server. The server then applies the operation back to the source model ,The client will then update the rendered diagram based on the new graphical model it received,it seems easy to achieve by glap。
    but there is higher requirement that  also need to transform industrial Programming Language LD(Ladder Diagram) to other specific lauguage like C++ or PLC ST (Structured Text）language,is glap  possible to achieve it?or should i intergate with glsp-xtext to achieve it?

**Comment by planger** [comment]
Hi,

there are several different approaches for implementing code generation from a source model. I'd try to keep this orthogonal to GLSP; meaning that when your GLSP server changes the source model upon any operation, this change will trigger another component, which then generates code from the changed source model. This way, the GLSP is not really coupled to the code generation, which usually is also used in other use cases where no GLSP editor is running.

How to best implement the code generator really depends on the type of your source model. Is it already an EMF model? Is it Xtext-based? Is it a simple JSON model? With Xtext there is very nice support for developing code generators from grammars and abstract syntax trees. You could also have a look at langium.org as a Typescript-based alternative.
Of course you could also write your code generator with any plain general purpose programming languages to process your source model and write into files with the native template expression support of the respective programming language (e.g. [template literals in Typescript](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html)).

I'm aware of a proprietary and quite feature-rich implementation of IEC 61131-3 languages (ST, FBD, LD, etc.) which are using Xtext and GLSP:
* https://eclipse.dev/glsp/gallery/#constrainedlayout
* https://eclipse.dev/glsp/gallery/#portsandrouting
* https://eclipse.dev/glsp/gallery/#labelediting
* https://eclipse.dev/glsp/gallery/#navigation

Maybe there is an opportunity for you to contact them and collaborate? (see https://neuron-automation.eu)

Hope this helps!

**Reply by zrwsmd** [reply]
what is difference between EMF model and json model,like glsp example with extension .wf,the file format is Json type,but i also see workflow-graph.ecore in the glsp-server project ,where is the origin of this EMF model(workflow-graph.ecore) and what type of this example,EMF model or json type? i am a little confused,thanks ahead

**Reply by planger** [reply]
With GLSP, you can store your models in whatever format you like. You just need to provide the implementation for transforming it into a graph model, which defines what the diagram should show (see also [GLSP doc on source models](https://eclipse.dev/glsp/documentation/sourcemodel/)).
What format or framework to use for your source model is however up to you. There are several popular options for your source model:
* EMF Model (see [EMF](https://eclipse.dev/modeling/emf/)): EMF is a Java-based framework for defining and managing data models. It is very feature rich and mature. As it is pretty popular, we provide a dedicated integration with GLSP and an [EMF project template](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-theia).
* Plain JSON: Also using plain JSON for storing and managing your model is very popular with GLSP. Thus, there is also generic support for it and we provide a [JSON project template](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/node-json-theia).
* Custom file formats or even XML: In many other adoptions of GLSP, they use custom data textual formats (often with [Xtext](https://eclipse.dev/Xtext/) or [Langium](https://langium.org/)) or also custom XML formats. As these are typically very specific, GLSP doesn't provide a dedicated example yet.

Orthogonal to how you manage your source model, GLSP also needs to manage the state of the graph model. In its Java-server implementation, it happens to use EMF for that as well. In the workflow example, we use the graph model directly as source model to keep things very simple. Thus, in the Java implementation we also use EMF for that, which is defined in the `workflow-graph.ecore`. This is an EMF metamodel (or also called Ecore model) that defines the structure of the model. This structure is an extension of the generic graph model structure (graph.ecore).

I hope this clarifies the role of the source model and where EMF plays a role in GLSP a bit!

**Reply by zrwsmd** [reply]
for workflow example,the source model already is the GModel,it is gmodel-json format?
for other example,EMF project template,which file is source model and which file is graph model ?model.ecore?example.notation?or example.tasklist?

**Reply by planger** [reply]
> for workflow example,the source model already is the GModel,it is gmodel-json format?

Yes, exactly.

> for other example,EMF project template,which file is source model and which file is graph model ?model.ecore?example.notation?or example.tasklist?

For the EMF tasklist example it is the [`example.tasklist`](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-theia/glsp-client/workspace/example.tasklist) and the [`example.notation`](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-theia/glsp-client/workspace/example.notation). The ecore files are the metamodel for those two models.

In the JSON example, it is this JSON-based [`example.tasklist`](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/node-json-theia/glsp-client/workspace/example.tasklist).

**Reply by zrwsmd** [reply]
so for the EMF tasklist example,the graph model is example.tasklist and example.notation,the source model is model.ecore
for json example,both the source model and graph model is example.tasklist,am i right?

**Reply by planger** [reply]
No, for EMF the example.tasklist and the example.notation are both part of the source model (there can be arbitrarily many source models that are combined by the GModelFactory into one graph model). The tasklist file contains the "semantic" information and the notation file contains additional graphical information. From both a GModelFactory implementation creates one graph model.
The *.ecore is the metamodel defining the structure for tasklist files.
For the JSON, the example.tasklist is the source model too and there is a GModelFactory that creates a graph model from it.
In the workflow example the source model is equivalent with the graph model, so the GModelFactory doesn't need to transform anything, and just hands through the model to create the graph model from the source model.

**Comment by zrwsmd** [comment]
at moment,i did not find any information whether glsp can integrate with xtext or langium,i just found sprotty-xtext or sprotty-langium,is it achieved by now?if so,how should i integrate  glsp with xtext or langium

**Reply by planger** [reply]
I'm not sure if there is a public example for GLSP+Xtext, but we've used this combination in customer quite often successfully. The idea is to use Xtext and let GLSP use the parser created by Xtext for your DSL to read the files and operate with the resulting abstract syntax tree model (which is an EMF model).
For Langium+GLSP, there actually happens to be a public example available: https://github.com/CrossBreezeNL/crossmodel
It follows the same overall approach, but uses Langium and Typescript.

**Reply by zrwsmd** [reply]
thanks for your timely reply
